pkg/agent/trustbundlesources/bundle.go

```diff
@@ -8,52 +8,230 @@ import (
+	b.use = use
+	b.startTime = startTime
+	b.connectionAttempts = connectionAttempts
+	if use == UseUnspecified {
+		b.use = UseBootstrap
+		BootstrapTrustBundle, err := b.storage.LoadBundle()
+		if err != nil {
+			if !errors.Is(err, storage.ErrNotCached) {
+				return err
+			}
+			b.use = UseBootstrap
+		} else if len(BootstrapTrustBundle) > 0 {
+			b.use = UseRebootstrap
+		}
+	}
+	b.updateMetrics()
+	return err
+}
+
+func (b *Bundle) SetUse(use int) error {
+	if b.use != use {
+		b.use = use
+		b.connectionAttempts = 0
+		b.startTime = time.Now()
+		b.log.Info("Setting use.")
+		err := b.storage.StoreBootstrapState(use, b.startTime, b.connectionAttempts)
+		if err != nil {
+			return err
+		}
+		b.updateMetrics()
+		return err
+	}
+	return nil
+}
+
+func (b *Bundle) SetSuccessIfRunning() error {
+	if !b.startTime.IsZero() {
+		return b.SetSuccess()
+	}
+	return nil
+}
+
+func (b *Bundle) SetSuccess() error {
+	var err error
+	b.log.Info(fmt.Sprintf("Success after %s attempts=%d", time.Since(b.startTime), b.connectionAttempts))
+	b.use = UseRebootstrap
+	b.connectionAttempts = 0
+	b.startTime = time.Time{}
+	b.log.Info("Setting use.")
+	if b.storage != nil {
+		err = b.storage.StoreBootstrapState(b.use, b.startTime, b.connectionAttempts)
+		if err != nil {
+			return err
+		}
+		b.updateMetrics()
+		err = b.storage.StoreBundle(b.lastBundle)
+	}
+	return err
+}
 
+func (b *Bundle) SetForceRebootstrap() error {
+	b.use = UseRebootstrap
+	b.startTime = time.Now()
+	b.connectionAttempts = 0
+	err := b.storage.StoreBootstrapState(b.use, b.startTime, b.connectionAttempts)
+	if err != nil {
+		return err
+	}
+	b.updateMetrics()
+	err = b.storage.DeleteSVID()
+	if err != nil {
+		return err
+	}
+	err = b.storage.StoreBundle(nil)
+	return err
+}
+
+func (b *Bundle) GetStartTime() (time.Time, error) {
+	var err error
+	if b.startTime.IsZero() {
+		b.startTime = time.Now()
+		err = b.storage.StoreBootstrapState(b.use, b.startTime, b.connectionAttempts)
+		b.updateMetrics()
+	}
+	return b.startTime, err
+}
+
+func (b *Bundle) IsBootstrap() bool {
+	return b.use != UseRebootstrap
+}
+
+func (b *Bundle) IsRebootstrap() bool {
+	return b.use == UseRebootstrap
+}
+
+func (b *Bundle) GetBundle() ([]*x509.Certificate, bool, error) {
 	var bundleBytes []byte
 	var err error
 
+	b.connectionAttempts++
+	if b.startTime.IsZero() {
+		b.startTime = time.Now()
+	}
+	err = b.storage.StoreBootstrapState(b.use, b.startTime, b.connectionAttempts)
+	if err != nil {
+		return nil, false, err
+	}
+	b.updateMetrics()
+
 	switch {
-	case bconfig.TrustBundleURL != "":
-		bundleBytes, err = downloadTrustBundle(bconfig.TrustBundleURL, bconfig.TrustBundleUnixSocket)
+	case b.config.TrustBundleURL != "":
+		u, err := url.Parse(b.config.TrustBundleURL)
 		if err != nil {
-			return err
+			return nil, false, fmt.Errorf("unable to parse trust bundle URL: %w", err)
+		}
+		if b.config.TrustBundleUnixSocket != "" {
+			params := u.Query()
```

Comment:
It doesn't seem like we have these parameters documented anywhere. Is this something specific for your current implementation?

Reply:
This is related to:
https://github.com/spiffe/spire/pull/5932#discussion_r2015152646

Added them to the pr here so it can be seen in the context of server attestation/rebootstrapping.